Save 10seconds when running NetworkSpec

pgrange opened the pull request:
Depending on how the different Ouroboros servers are instantiated during the tests, the tests can take 0.3 seconds or 10.3 seconds. This is due to [ipRetryDelay ](https://github.com/input-output-hk/ouroboros-network/blob/679c7da2079a5e9972a1c502b6a4d6af3eb76945/ouroboros-network-framework/src/Ouroboros/Network/Subscription/Worker.hs#L260).

We add a threadDelay before trying to connect to other peer so that we give our own listen thread to be already listening... which seems fair.

To check before merging:
* [x] CHANGELOG is up to date
* [x] Up to date with master

pgrange committed:
Save 10seconds when running NetwrokSpec

ch1bo commented on hydra-node/src/Hydra/Network/Ouroboros.hs:
```diff
@@ -143,6 +143,20 @@ withOuroborosNetwork tracer localHost remoteHosts networkCallback between = do
+    --       and the listen one, listening for other peers to connect.
+    --       When all nodes start at the same time, to be a good citizen, it
+    --       would be fair to first listen for the other peers to connect
+    --       before trying to connect to the other peers.
```
Can't we do this instead?

Looking at the `listen` code, it seems that the continuation there is invoked after starting to listen for connections. Would setting up the `listen`, then start the `connect` threads and only then `wait serverAsync` (see line 217) help this issue?

Anyhow. This is also fine the way it is.

pgrange replied:
Interesting. Let me take a look.

pgrange replied:
Thanks for your observation @ch1bo . I love the new code way better.

However, I'm not sure what are the consequences of having removed this instruction:
`wait serverAsync`

We'll see what the C.I. has to say about it but if you already know the answer, I'm interested.

ch1bo replied:
We likely need to pass the `serverAsync` to the continuation and call it still after start `connect`ing.

pgrange replied:
Actually, I don't get what's the purpose of the `wait serverAsync` instruction. Could you explain it?

pgrange replied:
`wait serverAsync` is essential as it ensures that, should the Ouroboros server thread crash for any exception, our code is aware of it and terminates its execution. Without it, we wouldn't notice that a thread has crashed and keep running but in an inconsistent state.

pgrange committed:
Use withServerNode callback to control connections

withServerNode last param is a callabck called when the server is
listening. We use it to control when we want to start connecting to
other nodes. Because we are good citizen and start by listening to
connection before trying to connect to peers ourselves.

pgrange committed:
FIX we wait for the Ouroboros server thread

By explicilitly waiting for the serverAsync we ensure that, in
cas this thread crashes for any reason, we acknowledge it and stop
our process.
